### PR TITLE
enterprise-portal: implement basic MSP IAM and RPCs

### PR DESCRIPTION
**Pull request by unknwon**
Closes CORE-99, closes CORE-176

This PR is based off (and was also served as PoC of) [RFC 962: MSP IAM framework](https://docs.google.com/document/d/1ItJlQnpR5AHbrfAholZqjH8-8dPF1iQcKh99gE6SSjs/edit). It comes with two main parts:

1. The initial version of the MSP IAM SDK: `lib/managedservicesplatform/iam`
	- Embeds the [OpenFGA server implementation](https://github.com/openfga/openfga/tree/main/pkg/server) and exposes the a `ClientV1` for interacting with it.
	- Automagically manages the both MSP IAM's and OpenFGA's database migrations upon initializing the `ClientV1`.
		CleanShot 2024-06-18 at 15 09 24@2x
	- Ensures the specified OpenFGA's store and automatization model DSL exists.
	- Utility types and helpers to avoid easy mistakes (i.e. make the relation tuples a bit more strongly-typed).
		- Decided to put all types and pre-defined values together to simulate a "central registry" and acting as a forcing function for services to form some sort of convention. Then when we migrate the OpenFGA server to a separate standalone service, it will be less headache about consolidating similar meaning types/relations but different string literals. 
1. The first use case of the MSP IAM: `cmd/enterprise-portal/internal/subscriptionsservice`
	- Added/updated RPCs:
		- Listing enterprise subscriptions via permissions
		- Update enterprise subscriptions to assign instance domains
		- Update enterprise subscriptions membership to assign roles (and permissions)
	- A database table for enterprise subscriptions, only storing the extra instance domains as Enterprise Portal is not the writeable-source-of-truth.

## Other minor changes

- Moved `internal/redislock` to `lib/redislock` to be used in MSP IAM SDK.
- Call `createdb ...` as part of `enterprise-portal` install script in `sg.config.yaml` (`msp_iam` database is a hard requirement of MSP IAM framework).

## Test plan

Tested with gRPC UI:

- `UpdateEnterpriseSubscription` to assign an instance domain
- `UpdateEnterpriseSubscriptionMembership` to assign roles
- `ListEnterpriseSubscriptions`:
	- List by subscription ID
	- List by instance domain
	- List by view cody analytics permissions

## After merge

- [ ] Roll out to dev and prod
- [ ] Delete previously created database `enterprise-portal` on dev and prod

### CODE REVIEW

File: cmd/enterprise-portal/internal/codyaccessservice/v1.go
Comment (unknwon) [inline cmd/enterprise-portal/internal/codyaccessservice/v1.go:1]: This file is changed only because of the change of signature for `samsm2m.RequireScope`.

File: lib/enterpriseportal/subscriptions/v1/subscriptions.proto [hunk 1]
```diff
@@ -215,16 +228,6 @@ message ListEnterpriseSubscriptionLicensesFilter {
...
-    // Return only product subscriptions whose license key contains the given member.
-    Member member = 5;
-  }
-
-  message Member {
-    // Return only product subscriptions contains the member with given SAMS
-    // account ID.
-    string sams_account_id = 1;
-    // Return only product subscriptions contains the member with the given permission.
-    string permission = 2;
```
Comment (unknwon) [inline lib/enterpriseportal/subscriptions/v1/subscriptions.proto:227]: Added to the wrong filter type last time 🥹 (this is for licenses not subscriptions)

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 2]
```diff
@@ -1,10 +1,164 @@
...
+	q := sqlf.Sprintf(`
+SELECT
+	id,
+	instance_domain
+FROM subscriptions
+WHERE %s
+%s`,
+		whereConds, limit,
+	)
+	rows, err := s.db.Query(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...)
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:86]: Curious: what's the advantage of using `sqlf` here, instead of just constructing the string and using `pgx` named query parameters?

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: I think the usage of named parameters looks like:

```go
Query(ctx, "select * from widgets where foo = @foo and bar = @bar", 
  pgx.NamedArgs{"foo": 1, "bar": 2}))
```

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: This feels a little more readable at the sites where we construct the queries, and avoids the noise of using `q.Query(sqlf.PostgresBindVar)` etc

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: Then, query constructors just:

1. Construct a query string with named parameters
2. At the same time, add to a `pgx.NamedArgs` set with the appropriate parameters

Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: "name query parameters" you mean "$1, $2" or "@where, @limit"? `sqlf` automatically computes the correct `X` in `$X`, so no manual counting (and could be source of semantic bug if args happen to have correct type even with 1-off).

File: lib/managedservicesplatform/iam/database.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+type metadata struct {
+	AuthorizationModelID  string    `gorm:"column:authorization_model_id"`
+	AuthorizationModelDSL string    `gorm:"column:authorization_model_dsl"`
+	UpdatedAt             time.Time `gorm:"not null"`
+}
+
+// migrateAndRecon migrates the "msp-iam" database schema (when needed) and
+// recons the framework metadata.
+func migrateAndRecon(ctx context.Context, logger log.Logger, sqlDB *sql.DB, redisClient *redis.Client) (_ *metadata, err error) {
+	ctx, span := iamTracer.Start(ctx, "iam.migrateAndRecon")
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/database.go:35]: Maybe this is just carry-over from my gamer days but "recon" strikes me as "reconnaissance" which I've never seen in a software context... I assume you mean "reconcile"? This is a pretty common term and not that much longer, I suggest we use the full word

```suggestion
// migrateAndReconcile migrates the "msp-iam" database schema (when needed) and
// recons the framework metadata.
func migrateAndReconile(ctx context.Context, logger log.Logger, sqlDB *sql.DB, redisClient *redis.Client) (_ *metadata, err error) {
	ctx, span := iamTracer.Start(ctx, "iam.migrateAndReconcile")
```

Comment (unknwon) [reply lib/managedservicesplatform/iam/database.go:35]: I do mean "reconnaissance" 😂 "Reconile" works here too I think.

File: lib/managedservicesplatform/iam/database.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	log.Logger
+}
+
+func (l *gooseLoggerShim) Fatalf(format string, v ...interface{}) {
+	l.Fatal("goose", zap.String("message", fmt.Sprintf(format, v...)))
+}
+
+func (l *gooseLoggerShim) Printf(format string, v ...interface{}) {
+	l.Debug("goose", zap.String("message", fmt.Sprintf(format, v...)))
+}
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/database.go:121]: nit: `"goose"` should be scope, let's just put the message in the message already formatted

```suggestion
type gooseLoggerShim struct {
	log.Logger
}

func (l *gooseLoggerShim) Fatalf(format string, v ...interface{}) {
	l.Fatal(fmt.Sprintf(format, v...))
}

func (l *gooseLoggerShim) Printf(format string, v ...interface{}) {
	l.Debug(fmt.Sprintf(format, v...))
}
```

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: I think constructing the parameters with `sqlf.Sprintf` is noisy as well. With named parameters you just append `condition = @var`. Then, converting `sqlf.Sprintf` back is also noisy (e.g. the line I commented on)

tl;dr I don't like `sqlf.Sprintf`, especially when native access to `pgx` gives us arguably better alternatives 😛

File: lib/managedservicesplatform/iam/database.go (hunk 3 above)
Comment (bobheadxi) [reply lib/managedservicesplatform/iam/database.go]: ah, or `interrogates`?

if it really is `recon` not a huge deal, but a bit confused myself 😆

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 2 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: Either named parameters (`condition = @var`) or bind parameters (`condition = $1`) requires maintain two list of args, and things become more loose when args are conditionally constructed/appended, which doesn't seem any more elegant then using `sqlf`.

> Then, converting sqlf.Sprintf back is also noisy (e.g. the line I commented on)

It is just matter of the common interface, the interface (aka. our thin wrapper around `pgxpool.Pool`) could have exposed to accept a `sqlf.Query` directly. 😀

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: One is a list of args, the other is _named_ args, which gives context on how each arg is used at the callsite. Lots of `foo = %s` IMO is pretty unreadable, especially the way we have to adapt string slices into `sqlf.Query` 😬

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:86]: Another way to think about it: compare `fmt.Sprintf` to JS/TS inline template strings, the latter is far better any day, we just have to deal with `fmt.Sprintf` because Go doesn't have nice inline template strings. Seems odd to have libraries that _try_ to use `fmt.Sprintf`-style patterns 😆

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 5]
```diff
@@ -1,10 +1,164 @@
...
+	}
+	return s.Get(ctx, subscriptionID)
+}
+
+// Get returns a subscription record with the given subscription ID. It returns
+// pgx.ErrNoRows if no such subscription exists.
+func (s *SubscriptionsStore) Get(ctx context.Context, subscriptionID string) (*Subscription, error) {
+	var subscription Subscription
+	q := sqlf.Sprintf(`SELECT id, instance_domain FROM subscriptions WHERE id = %s`, subscriptionID)
+	err := s.db.QueryRow(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...).Scan(&subscription.ID, &subscription.InstanceDomain)
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:159]:
```suggestion
	err := s.db.QueryRow(ctx, `SELECT id, instance_domain FROM subscriptions WHERE id = @id`,
		pgx.NamedArgs{"id": subscriptionID}).
		Scan(&subscription.ID, &subscription.InstanceDomain)
```

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:159]: (example for other thread of the cruft we can avoid from `sqlf.Sprintf`

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 6]
```diff
@@ -1,10 +1,164 @@
...
+	// ForceUpdate indicates whether to force update all fields of the subscription
+	// record.
+	ForceUpdate bool
+}
+
+func (opts UpsertSubscriptionOptions) toQuery(id string) (_ *sqlf.Query, noop bool) {
+	if opts.ForceUpdate {
+		return sqlf.Sprintf(`
+INSERT INTO subscriptions (id, instance_domain)
+VALUES (%s, %s)
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:114]: With native pgx:

```suggestion
INSERT INTO subscriptions (id, instance_domain)
VALUES (@instance_domain, @id)
```

then, we return `pgx.NamedArgs`

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:114]: (also, is this a bug? the parameters are currently ordered `opts.InstanceDomain, id`)

File: cmd/enterprise-portal/internal/dotcomdb/dotcomdb.go [hunk 7]
```diff
@@ -453,3 +459,37 @@ func (r *Reader) ListEnterpriseSubscriptionLicenses(
...
+// subscription IDs. The caller should check the length of the returned slice to
+// ensure all requested subscriptions were found.
+func (r *Reader) ListEnterpriseSubscriptions(ctx context.Context, subscriptionIDs ...string) ([]*SubscriptionAttributes, error) {
+	if len(subscriptionIDs) == 0 {
+		return []*SubscriptionAttributes{}, nil
+	}
+
+	idConds := sqlf.Join(dbutil.ArgsToQueries(subscriptionIDs), ", ")
+	q := sqlf.Sprintf("SELECT id, created_at, archived_at FROM subscriptions WHERE id IN (%s)", idConds)
+	rows, err := r.db.Query(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...)
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/dotcomdb/dotcomdb.go:480]: 85% sure you can do:

```suggestion
	rows, err := r.db.Query(ctx, "SELECT id, created_at, archived_at FROM subscriptions WHERE id ANY ($1)", subscriptionIDs)
```

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/dotcomdb/dotcomdb.go:480]: or a named arg if you'd like :-)

File: cmd/enterprise-portal/service/iam_model.fga [hunk 8]
```diff
@@ -0,0 +1,12 @@
+model
```
Comment (bobheadxi) [inline cmd/enterprise-portal/service/iam_model.fga:1]: Link to the FGA playground + how to edit this nicely there?

File: cmd/enterprise-portal/service/service.go [hunk 9]
```diff
@@ -152,7 +167,13 @@ func (Service) Initialize(ctx context.Context, logger log.Logger, contract runti
...
 		background.CallbackRoutine{
-			StopFunc: func(ctx context.Context) error {
+			StopFunc: func(context.Context) error {
+				closeIAMClient()
+				return nil
+			},
+		},
+		background.CallbackRoutine{
+			StopFunc: func(context.Context) error {
 				start := time.Now()
```
Comment (bobheadxi) [inline cmd/enterprise-portal/service/service.go:177]: Do these callback routines have names? Since there are multiple here we should name them IMO 🤔

File: cmd/enterprise-portal/internal/samsm2m/samsm2m.go [hunk 10]
```diff
@@ -27,10 +27,10 @@ func EnterprisePortalScope(permission scopes.Permission, action scopes.Action) s
 	return scopes.ToScope(scopes.ServiceEnterprisePortal, permission, action)
 }
 
-var tracer = otel.GetTracerProvider().Tracer("telemetry-gateway/samsm2m")
+var tracer = otel.GetTracerProvider().Tracer("enterprise-portal/samsm2m")
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/samsm2m/samsm2m.go:30]: 🤭

File: lib/managedservicesplatform/iam/server.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+type openfgaLoggerShim struct {
+	log.Logger
+}
+
+func (l *openfgaLoggerShim) Panic(s string, field ...zap.Field) {
+	l.Error(s, field...)
+}
+
+func (l *openfgaLoggerShim) DebugWithContext(_ context.Context, s string, field ...zap.Field) {
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/server.go:88]: Since we get context, maybe worth adding trace context to the log message as well? `internal/trace.Logger(ctx, logger).Debug(...)`

File: lib/managedservicesplatform/iam/server.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+func newServer(logger log.Logger, sqlDB *sql.DB) (*server.Server, error) {
+	openfgaLogger := &openfgaLoggerShim{Logger: logger}
+
+	// Defaults are all coming from github.com/openfga/internal/server/config/config.go,
+	// we cannot directly use them because of "internal" package.
+	datastore, err := postgres.NewWithDB(
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/server.go:26]: 😎

File: lib/managedservicesplatform/iam/iam.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+const (
+	TupleRelationMember TupleRelation = "member"
+	TupleRelationView   TupleRelation = "view"
+)
+
+type TupleSubject string
+
+// ToTupleSubjectUser creates a TupleSubject from a SAMS account ID for users,
+// e.g. "user:1234".
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/iam.go:38]: it's the external account ID, right? so `1234` might be a bit misleading (looks like internal numeric DB ID)

File: sg.config.yaml [hunk 14]
```diff
@@ -415,6 +415,8 @@ commands:
         export GCFLAGS='all=-N -l'
       fi
       go build -gcflags="$GCFLAGS" -o .bin/enterprise-portal github.com/sourcegraph/sourcegraph/cmd/enterprise-portal
+      # Ensure the "msp-iam" database exists (PostgreSQL has no "IF NOT EXISTS" option).
+      createdb -h $PGHOST -p $PGPORT -U $PGUSER msp-iam || true
```
Comment (bobheadxi) [inline sg.config.yaml:419]: 🙏

File: lib/redislock/mutex_test.go [hunk 15]
```diff
@@ -1,22 +1,26 @@
...
-	"github.com/sourcegraph/sourcegraph/internal/rcache"
 )
 
 func TestOnlyOne(t *testing.T) {
-	rcache.SetupForTest(t)
+	addr := os.ExpandEnv("$REDIS_HOST:$REDIS_PORT")
+	if addr == ":" {
+		t.Skip("REDIS_HOST and REDIS_PORT not set")
+	}
+	fmt.Println("addr:", addr)
```
Comment (bobheadxi) [inline lib/redislock/mutex_test.go:20]:
```suggestion
	t.Log("redis_addr", addr)
```

File: lib/managedservicesplatform/iam/database.go [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+	"go.opentelemetry.io/otel/codes"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+	gormlogger "gorm.io/gorm/logger"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/redislock"
+)
+
+const databaseName = "msp-iam"
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/database.go:23]: Since we have `enterprise_portal_subscriptions`, should this be `msp_iam`?

File: lib/managedservicesplatform/iam/server.go [hunk 17]
```diff
@@ -0,0 +1,215 @@
+package iam
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/server.go:1]: I recommend renaming `server.go` to `openfga_server.go`, for ease of discoverability of the OpenFGA component, and also to just reflect that this implementation embeds it

File: lib/managedservicesplatform/iam/server.go (hunk 12 above)
Comment (bobheadxi) [reply lib/managedservicesplatform/iam/server.go:26]: Did you need to fork OpenFGA for this?

Comment (bobheadxi) [reply lib/managedservicesplatform/iam/server.go:26]: Ah, found https://github.com/openfga/openfga/pull/1688 😁

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 2 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go]: Alright, updated in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/7bac5f3a89ea5e9383507a9a4c230b1e42dfcc49!

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 6 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go]: Yes it is a bug haha, fixed along with another comment.

File: cmd/enterprise-portal/service/iam_model.fga (hunk 8 above)
Comment (unknwon) [reply cmd/enterprise-portal/service/iam_model.fga:1]: I tried and FGA doesn't seem to support comments ... the best we can do is to add README?

Comment (unknwon) [reply cmd/enterprise-portal/service/iam_model.fga:1]: Actually... it is a bug in JetBrains plugin of OpenFGA 😠

Comment (unknwon) [reply cmd/enterprise-portal/service/iam_model.fga:3]: Added in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/9bbc4e0a7fc32a46f791c2d35be5e64b5cc5c986

File: cmd/enterprise-portal/service/service.go (hunk 9 above)
Comment (unknwon) [reply cmd/enterprise-portal/service/service.go]: They do, added in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/77fa71eecebd6b91145fc4910e4ac6da333c5446

File: lib/managedservicesplatform/iam/database.go (hunk 16 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/database.go:23]: `enterprise_portal_subscriptions` is a table name, following table naming convention... `msp_iam` is a database name, following DB name convention (or not)? 😂 we already have `enterprise-portal` database in MSP... should we fix that together (if possible?), or live with it forever?

`cody_management` did use underscore.

File: lib/redislock/mutex_test.go (hunk 15 above)
Comment (unknwon) [reply lib/redislock/mutex_test.go:20]: ah shit, this is debugging leftover 🤦‍♂️

File: lib/managedservicesplatform/iam/server.go (hunk 12 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/server.go:26]: Yep, tho not sure when would have time to finish that PR haha, I guess could be during my PTO, that's just a normal contribution to open source?

File: lib/managedservicesplatform/iam/server.go (hunk 11 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/server.go:88]: This file is under `lib/`... importing `internal/trace` would require moving `trace` to `lib/` as well, and from a quick skim that would create a huge cascading diff? I'd leave it out for now.

File: lib/managedservicesplatform/iam/iam.go (hunk 13 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/iam.go]: Updated in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/8e75be35df9b4b36e83af647d227519c94345cbe

File: lib/managedservicesplatform/iam/database.go (hunk 16 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/database.go]: renamed in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/4330b8f40df8db10b9df4e6ea741eb71596edd34

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go]: nice, thank you!

File: lib/managedservicesplatform/iam/database.go (hunk 16 above)
Comment (bobheadxi) [reply lib/managedservicesplatform/iam/database.go]: 🤯

File: lib/managedservicesplatform/iam/server.go (hunk 11 above)
Comment (bobheadxi) [reply lib/managedservicesplatform/iam/server.go]: Ah, you're right. You can just use the native `logger.WithTrace` here then, and make a small one-off helper?

Comment (bobheadxi) [reply lib/managedservicesplatform/iam/server.go]: I think having a trace context would help put the third party logs into context, which we may be glad to have when it comes down to it

File: cmd/enterprise-portal/internal/codyaccessservice/v1_store.go [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+type storeV1 struct {
+	SAMSClient *sams.ClientV1
+}
+
+type NewStoreV1Options struct {
+	SAMSClient *sams.ClientV1
+}
+
+// NewStoreV1 returns a new StoreV1 using the given resource handles.
+func NewStoreV1(opts NewStoreV1Options) StoreV1 {
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/codyaccessservice/v1_store.go:30]: nit: I think the convention would be a simpler:
```suggestion
type StoreV1Options struct {
	SAMSClient *sams.ClientV1
}

// NewStoreV1 returns a new StoreV1 using the given resource handles.
func NewStoreV1(opts StoreV1Options) StoreV1 {
```

File: lib/managedservicesplatform/iam/database.go [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+	"go.opentelemetry.io/otel/codes"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+	gormlogger "gorm.io/gorm/logger"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/redislock"
+)
+
+const databaseName = "msp_iam"
```
Comment (bobheadxi) [inline lib/managedservicesplatform/iam/database.go:23]: sorry for the flip-flop but maybe we _should_ align on "-" for database name, I didn't realize we already have db `enterprise-portal` 😅 But not a huge deal

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 20]
```diff
@@ -1,10 +1,161 @@
...
+	// PageSize is the maximum number of subscriptions to return.
+	PageSize int
+}
+
+func (opts ListEnterpriseSubscriptionsOptions) toQueryConditions() (where, limit string, _ pgx.NamedArgs) {
+	whereConds := []string{"TRUE"}
+	namedArgs := pgx.NamedArgs{}
+	if len(opts.IDs) > 0 {
+		whereConds = append(whereConds, "id = ANY(@ids)")
+		namedArgs["ids"] = opts.IDs
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:56]: 💯

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 21]
```diff
@@ -1,10 +1,161 @@
...
+
+// toQuery returns the query based on the options. It returns an empty query if
+// nothing to update.
+func (opts UpsertSubscriptionOptions) toQuery(id string) (query string, _ pgx.NamedArgs) {
+	const queryFmt = `
+INSERT INTO enterprise_portal_subscriptions (id, instance_domain)
+VALUES (@id, @instanceDomain)
+ON CONFLICT (id)
+DO UPDATE SET
+	%s`
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:118]: nit: might be easier if we do something like:

```go
query += fmt.Sprintf(`DO UPDATE SET %s`, sets) 
```

later around L131 instead. Feels tricky having to follow a `%s` to the final "format", and what if a code path is introduced that returns early?

File: lib/enterpriseportal/subscriptions/v1/subscriptions.proto [hunk 22]
```diff
@@ -200,6 +198,17 @@ enum EnterpriseSubscriptionLicenseType {
   ENTERPRISE_SUBSCRIPTION_LICENSE_TYPE_KEY = 1;
 }
 
+message ListEnterpriseSubscriptionsFilter {
+  oneof filter {
+    // Return only product subscriptions has the given subscription ID.
+    string subscription_id = 1;
+    // Return only product subscriptions with the given archival status.
+    bool is_archived = 2;
+    // Return only product subscriptions satisfies the given permission.
```
Comment (bobheadxi) [inline lib/enterpriseportal/subscriptions/v1/subscriptions.proto:207]:
```suggestion
    // Return only product subscriptions that satisfy the given permission.
```

File: lib/enterpriseportal/subscriptions/v1/subscriptions.proto [hunk 23]
```diff
@@ -342,20 +344,45 @@ message EnterpriseSubscriptionMembership {
...
+  // The type of the permission.
+  PermissionType type = 1;
+  // The relation between the user and the resource.
+  PermissionRelation relation = 2;
+  // The SAMS account ID of the user.
+  string sams_account_id = 3;
+}
+
+// PermissionType represents the type of permissions.
+enum PermissionType {
```
Comment (bobheadxi) [inline lib/enterpriseportal/subscriptions/v1/subscriptions.proto:379]: Is it accurate to say that "type of permission" == "entity" or "object"? If we can describe it as such it might be clearer to someone new to the model

**Review by bobheadxi — APPROVED** [review]
Overall LGTM, super exciting! A few comments and a [question](https://github.com/sourcegraph/sourcegraph/pull/63173#issuecomment-2179318250)

File: cmd/enterprise-portal/internal/database/subscriptions.go [hunk 24]
```diff
@@ -1,10 +1,161 @@
...
+)
+
 // Subscription is a product subscription record.
 type Subscription struct {
 	// ID is the prefixed UUID-format identifier for the subscription.
 	ID string `gorm:"primaryKey"`
 	// InstanceDomain is the instance domain associated with the subscription, e.g.
 	// "acme.sourcegraphcloud.com".
-	InstanceDomain string `gorm:"index"`
+	InstanceDomain string `gorm:"unique"`
```
Comment (bobheadxi) [inline cmd/enterprise-portal/internal/database/subscriptions.go:20]: What happens with archived subscriptions? 🤔 where a subscription is archived, but then a customer comes back and uses the same domain

For implementing a full table, should I have a separate table, `archived_enterprise_portal_subscriptions`? 🤔

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:20]: But, our list RPCs offer filters for archived vs not, so it should remain accessible even after archive

Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions.go:20]: Maybe we just expect Update to still work on archived subscriptions, and if there's a conflict, just update the old archived subscription?

File: lib/managedservicesplatform/iam/server.go (hunk 11 above)
Comment (unknwon) [reply lib/managedservicesplatform/iam/server.go]: Added in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/4292a0208451a53821c4141a7d9d1a79b3a789ba

File: cmd/enterprise-portal/internal/codyaccessservice/v1_store.go (hunk 18 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/codyaccessservice/v1_store.go]: Updated in https://github.com/sourcegraph/sourcegraph/pull/63173/commits/ce3b9e81d5de04c41864d15ad4d139d4327ad0dc

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 21 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go:119]: When that happen, tests should catch it (or the change maker should own refactoring) 😁 I am trying to avoid string concatenation as much as possible because that makes digest the full picture of the query harder.

**Comment by unknwon** [comment]
> General question: we return prefixed UUIDs for external consumption, are we reconciling that with our internal IDs? e.g. internally do we use non-prefixed, and only add the prefix at the RPC layer? In that case, should we trim prefixes at the RPC layer? Allow prefixed and non-prefixed?

I have no preference here, unless you have strong opinion about "let's take this chance to prefix all subscription IDs with `es_`", but again all over the places are expecting a UUID (more or less, don't know if some places even cares about the length, then the change of length could be headache).

I am leaning towards only keeping it at RPC layer, which is the strategy this PR applies:

CleanShot 2024-06-19 at 20 08 34@2x

or maybe get rid of the prefix thing for subscription IDs altogether. When we introduce a new type of ID, we make sure we prefix from the beginning (like the case for SAMS account ID that did not have a prefix... we live with it and prefix all other resource IDs).

What's your take here?

File: cmd/enterprise-portal/internal/database/subscriptions.go (hunk 24 above)
Comment (unknwon) [reply cmd/enterprise-portal/internal/database/subscriptions.go:20]: Hmm, good call, ideally the unique condition should be `where archived_at is null` (similar to [exclude soft-deleted records](https://github.com/sourcegraph/sourcegraph-accounts/blob/bfdbf93fcc4df74a848c100d77b70f34a12e4df7/backend/internal/database/external_account_connections.go#L42)), but we don't have the full table schema yet. 

We can change the unique condition when we introduce the `archived_at` field? (IIRC GORM will take care of updating it, even if not, we only need to drop the current unique index right before rolling out the new version, GORM will then create the correct unique index).

File: lib/enterpriseportal/subscriptions/v1/subscriptions.proto (hunk 23 above)
Comment (bobheadxi) [reply lib/enterpriseportal/subscriptions/v1/subscriptions.proto:413]: I think in-code we should align on terms, but docstrings we should try to help people understand by throwing lots of terms out

Comment (bobheadxi) [reply lib/enterpriseportal/subscriptions/v1/subscriptions.proto:413]:
```suggestion
// PermissionType defines a class of objects with similar characteristics.
// https://openfga.dev/docs/concepts#what-is-a-type
enum PermissionType {
```

**Comment by bobheadxi** [comment]
> What's your take here?

Let's do the RPC layer strategy, and keep it prefix-free internally

**Comment by unknwon** [comment]
Merging as-is, happy to address post-merge feedback!
